File: lab5/c/dense1.c
#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#include "time.h"

#define IDX(i, j, n) ((i) * ((i) + 1) / 2 + (j)) // <- OPT 2

#define BLKSIZE 8

static int
max(int a, int b)
{
    if (a > b)
        return (a);
    else
        return (b);
}

int chol(double *A, const unsigned int n)
{
    register unsigned int i;
    register unsigned int j;
    register unsigned int k;

    register double tmp;

    for (j = 0; j < n; j++)
    {
        for (i = j; i < n; i++)
        {
            tmp = A[IDX(i, j, n)]; /* <- OPT 1 */
            for (k = 0; k < j;)
            {
                if (k < max(j - BLKSIZE, 0))
                { // <- OPT 3
                    tmp -= A[IDX(i, k, n)] * A[IDX(j, k, n)] +
                           A[IDX(i, k + 1, n)] * A[IDX(j, k + 1, n)] +
                           A[IDX(i, k + 2, n)] * A[IDX(j, k + 2, n)] +
                           A[IDX(i, k + 3, n)] * A[IDX(j, k + 3, n)] +
                           A[IDX(i, k + 4, n)] * A[IDX(j, k + 4, n)] +
                           A[IDX(i, k + 5, n)] * A[IDX(j, k + 5, n)] +
                           A[IDX(i, k + 6, n)] * A[IDX(j, k + 6, n)] +
                           A[IDX(i, k + 7, n)] * A[IDX(j, k + 7, n)];
                    k += BLKSIZE;
                }
                else
                {
                    tmp -= A[IDX(i, k, n)] *
                           A[IDX(j, k, n)];
                    k++;
                }
            }
            A[IDX(i, j, n)] = tmp;
        }

        if (A[IDX(j, j, n)] < 0.0)
            return (1);

        tmp = A[IDX(j, j, n)] = sqrt(A[IDX(j, j, n)]);
        for (i = j + 1; i < n; i++)
        {
            A[IDX(i, j, n)] /= tmp;
        }
    }

    return (0);
}

double *
alloc_matrix(int n)
{
    double *ret;

    ret = calloc(n * (n + 1) / 2, sizeof(double));
    assert(ret != NULL);

    return (ret);
}

int main(int argc, char **argv)
{
    double *A;
    double t1, t2;
    int i, j, n, ret;

    n = atoi(argv[1]);
    A = alloc_matrix(n);

    for (i = 0; i < n; i++)
        A[IDX(i, i, n)] = 1.0;

    t1 = stamp();
    if (chol(A, n))
        fprintf(stderr, "Error: matrix is either not symmetric or not positive definite.\n");
    t2 = stamp();
    fprintf(stdout, "C:\t%lg [s]\n", t2 - t1);
    free(A);
    return 0;
}

File: lab5/go/dense1.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"time"
)

func makeTimestamp() int64 {
	return time.Now().UnixNano()
}

func IDX(i uint, j uint, n uint) uint {
	return j + i*n
}

func chol(A []int /*OPT*/, n uint) int {
	var j uint
	var i uint
	var k uint
	var tmp int /*OPT*/

	for j = 0; j < n; j++ {
		for i = j; i < n; i++ {
			tmp = A[IDX(i, j, n)]
			for k = 0; k < j; k++ {
				tmp = tmp - (A[IDX(i, k, n)] * A[IDX(j, k, n)])
			}
			A[IDX(i, j, n)] = tmp
		}
		var real float64 = float64(A[IDX(j, j, n)])
		if real < 0.0 {
			return (1)
		}
		real = math.Sqrt(real)
		A[IDX(j, j, n)] = int(real)
		for i = j + 1; i < n; i++ {
			A[IDX(i, j, n)] = int(float64(A[IDX(i, j, n)]) / real)
		}
	}
	return (0)
}

func show(A []int, n uint) {
	var i uint
	var j uint
	for i = 0; i < n; i++ {
		for j = 0; j < n; j++ {
			fmt.Print(A[IDX(i, j, n)])
		}
		fmt.Println()
	}
}

func main() {
	nx, _ := strconv.Atoi(os.Args[1])
	var n uint = uint(nx)
	var t1 int64
	var t2 int64
	var A []int = make([]int, n*n) /*OPT*/

	var i uint
	for i = 0; i < n-1; i++ {
		A[IDX(i, i, n)] = 2.0
		A[IDX(i, i+1, n)] = 1.0
		A[IDX(i+1, i, n)] = 1.0
	}
	A[IDX(n-1, n-1, n)] = 2.0

	if len(os.Args) > 2 {
		show(A, n)
	}

	t1 = makeTimestamp()
	if chol(A, n) != 0 {
		fmt.Println("error")
	}
	t2 = makeTimestamp()

	if len(os.Args) > 2 {
		show(A, n)
	}

	fmt.Print("GO:\t")
	fmt.Print(float64(t2-t1) / float64(1000000000))
	fmt.Println(" [s]")
}
